Read options from a manifest file in importcontent

dylanmccall opened the pull request:
## Summary

This adds a `--manifest` option to the `importcontent` management command.

With this change, one can export content from one Kolibri instance…

```
kolibri manage exportcontent c9d7f950ab6b5a1199e3d6c10d7f0103 my_export_dir
```

And then import it in another Kolibri instance…

```
kolbri manage importchannel network c9d7f950ab6b5a1199e3d6c10d7f0103
kolibri manage importcontent --manifest=my_export_dir/content/manifest.json disk c9d7f950ab6b5a1199e3d6c10d7f0103
```

Or…

```
kolbri manage importchannel network c9d7f950ab6b5a1199e3d6c10d7f0103
kolibri manage importcontent disk c9d7f950ab6b5a1199e3d6c10d7f0103 my_export_dir
```

That second option changes some existing behaviour: it automatically detects if there is a file named `manifest.json` in the path we are importing from, and uses _that_ for the default `node_ids` and `exclude_node_ids` (instead of the previous behaviour where it would attempt to import everything). This should be the same in practice, as `manifest.json` should select all of the content which has been exported. I add a `--no_detect_manifest` option to preserve the previous behaviour.

## References

This is a counterpart to my changes in #9460, adding support for _exporting_ a manifest file.

## Reviewer guidance

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

dylanmccall committed:
Read options from a manifest file in importcontent

dylanmccall committed:
Detect manifest files in importcontent

danigm commented on kolibri/core/content/utils/import_export_content.py:
```diff
@@ -185,7 +185,7 @@ def get_import_export_nodes(  # noqa: C901
 
         # filter down the query to remove files associated with nodes we've specifically been asked to exclude
-        if exclude_node_ids:
+        if exclude_node_ids is not None:
```
The `--node_ids` and `--exclude_node_ids` defaults to `[]` in the `exportcontent.py` and `importcontent.py` commands, so it will never be `None` when calling from there. I don't know if I'm missing something, but when I test the export command with `exclude_node_ids` option, I don't get anything, because the `nodes_query` is empty, because in the previous `if`, the filter is done with an empty `node_ids` list.

dylanmccall replied:
Oh yeah, this commit seems rather suspect. Thanks for pointing it out! If we want to select a whole channel, `exportcontent` should just set `node_ids` to the channel ID (the root node), so this case where node_ids could be None doesn't make sense and it's better if we treat None and `[]` as equivalent :)

I'll revert it, and add some unit tests to make sure things work as needed.

dylanmccall replied:
Never mind, this made sense in the end. I did need to change some tests, but this ended up a useful distinction for the case where a manifest file needs to indicate that there is no content. Just needed a bit more documentation :)

So what's happening here is, if you run `importcontent` without specifying any node_ids, this function gets called with `node_ids=None`. That does the right thing and matches with all nodes in the channel. There was some funny behaviour before, where if you do `importcontent --node_ids=''`, this function ends up being called with `node_ids=['']`, which also does the right thing (it matches with no nodes in the channel), but in the wrong way (it only matches because after an exhaustive search we have concluded there's no node with ID `''`). But that's great, because it means cleaning this up so `None` and `[]` are different doesn't change the behaviour!

So in the end I didn't revert this, but I added an explanatory comment in `get_import_export_nodes`, and I added some unit tests for the specific cases where `node_ids=[]` and `node_ids=None`.

danigm commented on kolibri/core/content/management/commands/importcontent.py:
```diff
@@ -565,6 +630,16 @@ def _start_file_transfer(self, f, filetransfer):
         return FILE_TRANSFERRED, data_transferred
 
     def handle_async(self, *args, **options):
+        if not options["directory"] and not options["manifest"]:
+            raise CommandError(
+                "Either a directory or a manifest file must be provided."
+            )
```
This makes impossible to import from network without a manifest. And there's an exception because `options["directory"]` is not an existing option when importing from network

dylanmccall replied:
Good spotting! That's fixed in the newest push.

dylanmccall committed:
Handle empty node_ids in importcontent

If node_ids is set to an empty string, read it as an empty list,
instead of a list with a single empty string.

rtibbles commented on kolibri/core/content/management/commands/importcontent.py:
```diff
@@ -61,6 +63,24 @@ def add_arguments(self, parser):
+        """
+        parser.add_argument(
+            "--manifest",
+            # Split the comma separated string we get, into a list of strings
```
This code comment seems like it was a copy paste from below?

rtibbles reviewed (changes requested):
I don't see any issues on the face of the code, but it would be good to add some test coverage of this new functionality.

dylanmccall committed:
Support manifest_file without a name attribute

This could be a file-like object extended from IOBase.

dylanmccall committed:
Clean up repeated node ID in ImportContentTestCase

dylanmccall committed:
Add tests for network import with manifest files

dylanmccall committed:
Allow empty node_ids to override a manifest file

In importcontent, node_ids is None needs to be treated differently
from node_ids=[].

dylanmccall committed:
Add tests for local import with manifest files

dylanmccall commented:
Okay, I added some unit tests here, both for `importcontent remote` and `importcontent disk` with manifest files. I think they cover all the cases I've added here, including the `--no-detect-manifest` option and manifest files with mismatched channel versions.

dylanmccall committed:
Use six.StringIO instead of io.StringIO in tests

This resolves a compatibility issue with Python 2.

dylanmccall committed:
Fix Py2 unicode errors in ImportContentTestCase

Node IDs turn into unicode strings when read from the database, so
we need to explicitly check for unicode strings in assertions from
get_import_export_mock.

dylanmccall committed:
Document new tests in ImportContentTestCase

rtibbles commented on kolibri/core/content/management/commands/importcontent.py:
```diff
@@ -71,8 +90,8 @@ def add_arguments(self, parser):
-            # Split the comma separated string we get, into a list of strings
-            type=lambda x: x.split(","),
+            # Split the comma separated string we get, into a set of strings
+            type=lambda x: set(x.split(",")) if x else set(),
```
Could we do this coersion to a set inside the `handle_async` method instead of here? Currently this introduces a different in behaviour between command line invocations of the management command and calls to `call_command` - as you can see below in the tests you didn't modify, `node_ids` and `exclude_node_ids` are still being set as lists. If the content manifest behaviour in particular is sensitive to this, then this may cause issues.

dylanmccall replied:
Yeah, sure thing! In the newest commit, the argument is parsed into a list (an empty list if it is given an empty string), and we coerce it into a set inside `_transfer()`. I'm not 100% sure about doing it there, but it kind of felt right since we clean up some other inputs there as well.

dylanmccall committed:
In importcontent, use a set for node_ids

This makes the code easier to test, and better reflects that the order
of node_ids is not significant. Related code in content_manifest already
deals with node_ids as a set.
